Zend/Optimizer/sccp.c

```diff
@@ -800,57 +800,115 @@ static inline zend_result ct_eval_array_key_exists(zval *result, zval *op1, zval
 }
 
 static bool can_ct_eval_func_call(zend_string *name, uint32_t num_args, zval **args) {
+#define CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(len, cond) \
+	case len: { \
+		if ((cond)) { return true; } \
+		break; \
+	}
+
 	/* Functions in this list must always produce the same result for the same arguments,
 	 * and have no dependence on global state (such as locales). It is okay if they throw
-	 * or warn on invalid arguments, as we detect this and will discard the evaluation result. */
-	if (false
-		|| zend_string_equals_literal(name, "array_diff")
-		|| zend_string_equals_literal(name, "array_diff_assoc")
-		|| zend_string_equals_literal(name, "array_diff_key")
-		|| zend_string_equals_literal(name, "array_flip")
-		|| zend_string_equals_literal(name, "array_is_list")
-		|| zend_string_equals_literal(name, "array_key_exists")
-		|| zend_string_equals_literal(name, "array_keys")
-		|| zend_string_equals_literal(name, "array_merge")
-		|| zend_string_equals_literal(name, "array_merge_recursive")
-		|| zend_string_equals_literal(name, "array_replace")
-		|| zend_string_equals_literal(name, "array_replace_recursive")
-		|| zend_string_equals_literal(name, "array_unique")
-		|| zend_string_equals_literal(name, "array_values")
-		|| zend_string_equals_literal(name, "base64_decode")
-		|| zend_string_equals_literal(name, "base64_encode")
+	 * or warn on invalid arguments, as we detect this and will discard the evaluation result.
+	 *
+	 * In PHP 8.2, many functions stopped depending on locales due to https://wiki.php.net/rfc/strtolower-ascii
+	 *
+	 * Break up a really long if-else chain that the compiler can't optimize by switching on the length for the following reasons:
+	 * 1. Improve worst-case performance.
+	 * 2. Make it less of a concern whether a function is used often enough to include in this list. */
+	switch (ZSTR_LEN(name)) {
+		CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(3, false
+			|| zend_string_equals_literal(name, "pow")
+		);
+		CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(4, false
+			|| zend_string_equals_literal(name, "trim")
+		);
+		CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(5, false
+			|| zend_string_equals_literal(name, "crc32")
+			|| zend_string_equals_literal(name, "ltrim")
+			|| zend_string_equals_literal(name, "rtrim")
+		);
+		CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(6, false
+			|| zend_string_equals_literal(name, "strcmp")
+			|| zend_string_equals_literal(name, "strpos")
+			|| zend_string_equals_literal(name, "strstr")
+			|| zend_string_equals_literal(name, "substr")
+		);
+		CAN_CT_EVAL_FUNC_CALL_CHECK_CASE(7, false
 #ifndef ZEND_WIN32
-		/* On Windows this function may be code page dependent. */
-		|| zend_string_equals_literal(name, "dirname")
+			/* On Windows this function may be code page dependent. */
+			|| zend_string_equals_literal(name, "dirname")
 #endif
```

Comment:
This causes [compile issues on Windows](https://ci.appveyor.com/project/php/php-src/builds/41902659/job/99623dem2q2hgl8q#L594). Apparently, the macro is evaluated before the preprocessing directive. That might be different with the [new MSVC preprocessor](https://docs.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-170).